cirq/contrib/two_qubit_gates/math_utils.py

```diff
@@ -0,0 +1,292 @@
+        are (broadcastable) arraylike objects whose sum would have shape (...),
+        the output is an array of shape (...,2,2), where the final two indices
+        correspond to unitary matrices.
+    """
+
+    U00 = np.cos(theta) * np.exp(1j * np.asarray(phi_d))
+    U10 = 1j * np.sin(theta) * np.exp(1j * np.asarray(phi_o))
+
+    # This implementation is agnostic to the shapes of the angles, as long
+    # as they can be broadcast together.
+    Udiag = np.array([[U00, np.zeros_like(U00)],
+                      [np.zeros_like(U00), U00.conj()]])
+    Udiag = np.moveaxis(Udiag, [0, 1], [-2, -1])
+    Uoff = np.array([[np.zeros_like(U10), -U10.conj()],
+                     [U10, np.zeros_like(U10)]])
+    Uoff = np.moveaxis(Uoff, [0, 1], [-2, -1])
+    return Udiag + Uoff
+
+
+def random_qubit_unitary(number: int = 1,
+                         sample_phase: bool = False) -> np.ndarray:
+    """Random qubit unitary distributed over the Haar measure.
+
+    The implementation is vectorized for speed.
+
+    Args:
+        number: If not 1, an ndarray of shape (number,2,2) is returned.
+            Otherwise a single 2x2 matrix is returned.
+        sample_phase: (Default False) If True, a global phase is also sampled
+            randomly. This corresponds to sampling over U(2) instead of SU(2).
+    """
+    theta = np.arcsin(np.sqrt(np.random.rand(number)))
+    phi_d = np.random.rand(number) * TWO_PI
+    phi_o = np.random.rand(number) * TWO_PI
+
+    out = _single_qubit_unitary(theta, phi_d, phi_o)
+
+    if sample_phase:
+        # coverage: ignore
+        global_phase = np.exp(1j * TWO_PI * np.random.rand(number))
+        np.einsum('t,tab->tab', global_phase, out, out=out)
+    return out
+
+
+def vector_kron(first: np.ndarray, second: np.ndarray) -> np.ndarray:
+    """Vectorized implementation of kron for square matrices."""
+    s_0, s_1 = first.shape[-2:], second.shape[-2:]
+    assert s_0[0] == s_0[1]
+    assert s_1[0] == s_1[1]
+    out = np.einsum('...ab,...cd->...acbd', first, second)
+    s_v = out.shape[:-4]
+    return out.reshape(s_v + (s_0[0] * s_1[0],) * 2)
+
+
+_MAGIC = np.array([[1, 0, 0, 1j], [0, 1j, 1, 0], [0, 1j, -1, 0], [1, 0, 0, -1j]
+                  ]) * np.sqrt(0.5)
+
+_gamma = np.array([[1, 1, 1, 1], [1, 1, -1, -1], [-1, 1, -1, 1], [1, -1, -1, 1]
+                  ]) * 0.25
+
+# all permutations of (1,2,3)
+_perms_123 = np.zeros((6, 3, 3), int)
+for ind, perm in enumerate(permutations((0, 1, 2))):
+    _perms_123[ind, (0, 1, 2), perm] = 1
+
+_negations = np.zeros((4, 3, 3), int)
+_negations[0, (0, 1, 2), (0, 1, 2)] = 1
+_negations[1, (0, 1, 2), (0, 1, 2)] = (1, -1, -1)
+_negations[2, (0, 1, 2), (0, 1, 2)] = (-1, 1, -1)
+_negations[3, (0, 1, 2), (0, 1, 2)] = (-1, -1, 1)
+
+_offsets = np.zeros((4, 3))
+_offsets[1, (1, 2)] = np.pi / 2
+_offsets[1, (0, 2)] = np.pi / 2
+_offsets[1, (0, 1)] = np.pi / 2
+
+
+def _kak_equivalent_vectors(kak_vec) -> np.ndarray:
```

Comment:
Instead of broadcasting single kak vectors into many possible vectors, you should canonicalize incoming vectors into their canonical form. This will reduce space usage.

Reply:
I think we do need to consider equivalent vectors, specifically if we are near the k_x = pi/4 surface of the Weyl chamber.

```python
x, y, z = np.pi / 4, 1, 0.5
kak_0 = kak_canonicalize_vector(x, y, z).interaction_coefficients
kak_1 = kak_canonicalize_vector(x - 1e-3, y, z).interaction_coefficients

inf_check_equivalent = KAK_vector_infidelity(kak_0, kak_1, False)
inf_ignore_equivalent = KAK_vector_infidelity(kak_0, kak_1, True)

print(f'infidelity without checking equivalent: {inf_ignore_equivalent}')
print(f'infidelity with checking equivalent: {inf_check_equivalent}')
```
returns:
```
infidelity without checking equivalent: 0.7080737102000556
infidelity with checking equivalent: 9.999996666287814e-07
```

Checking this actually revealed me an error in the code. I should probably write some tests to confirm that `KAK_vector_infidelity` is working properly.  In practice I am only using it with the `ignore_equivalent_vectors=True` option, in which case it only gives an upper bound to the infidelity.